feat: add method to compare graphs

Kirill89 opened the pull request:
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

Add method to compare graphs (`equals()`).

adrukh commented:
Consider even loop spinning as part of this logic, could be time-consuming?

Kirill89 commented:
@adrukh, I'm not sure - don't think we will have too many iterations here.
@michael-go what do you say?

Kirill89 commented:
> Tests!

👓

michael-go commented on test/core/is-same.test.ts:
```diff
@@ -0,0 +1,21 @@
+    expect(simpleGraphB.isSame(simpleGraphA)).toBe(true);
+  });
+
+  test('compare different graphs', async () => {
```
please add another test where the graphs are only slightly different - in example same packages but different relations

Kirill89 replied:
done

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +138,38 @@ class DepGraphImpl implements types.DepGraphInternal {
+  public isSame(other: types.DepGraph): boolean {
+    const sortByNodeId = (nodeA: any, nodeB: any) => nodeA.nodeId.localeCompare(nodeB.nodeId);
+
+    const selfNodes = this.toJSON().graph.nodes.sort(sortByNodeId);
```
you don't have to do this, you can cast `other` to `DepGraphImpl` and compare the privates ... maybe it's better to avoid some wasted CPU cycles?

Kirill89 replied:
I agree, but `DepGraph` is the only public interface here. I can't be sure `other` is the `DepGraphImpl` implementation.

michael-go replied:
I think you can assume it. We already do something similar in https://github.com/snyk/dep-graph/blob/master/src/legacy/index.ts#L203

gjvis replied:
I don't think this is a great idea, its ok in "legacy conversion" as we want that out as fast as possible, but we can't guarantee that the runtime types are what we expect can we?

gjvis replied:
if we need to do this for efficiency reasons than maybe that's telling us something about the interface itself? TS is not JS

michael-go replied:
@gjvis you have a point. In theory 2 different versions of the lib can exist in the same app and 2 incompatible instances can "meet" in this comparison function.

Doing 
```typescriprt
fromJSON(other.toJSON()) 
```
shouldn't be expensive, and is indeed safer.

and yep it might hint of a better potential interface 🤔 ... can't see it right now.

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +138,38 @@ class DepGraphImpl implements types.DepGraphInternal {
+    }
+
+    for (let i = 0; i < selfNodes.length; i++) {
+      if (selfNodes[i].nodeId !== otherNodes[i].nodeId) {
```
please compare not only the nodeIds, but also the data on each node, specifically the `pkgId` & `nodeInfo`

Kirill89 replied:
Why? Can you explain please?

michael-go replied:
The data the graph holds is:
* the relations between the *packages* (every node points to a `pkgId`)
* optional per node metadata
If one of these changes, than it's basically a different dep-graph

*actually* the `nodeIds` themselves don't matter at all, what matters is what `pkgId` they represent.

So the perfect comparison will ignore `nodeIds` and will compare the actual relations. This is not a must if this complicates the solution too much right now and blocks you, as usually the dep-graphs will be constructed in the same way so the `nodeId` of same graphs *should* be the same, but it's not always the case ...

Kirill89 replied:
done

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +138,38 @@ class DepGraphImpl implements types.DepGraphInternal {
   }
 
+  public isSame(other: types.DepGraph): boolean {
+    const sortByNodeId = (nodeA: any, nodeB: any) => nodeA.nodeId.localeCompare(nodeB.nodeId);
```
let's also compare the packages (and the the `pkgInfo` on each)

Kirill89 replied:
done

Kirill89 commented:
@garethr , @michael-go please re-review.

garethr commented:
@Kirill89 looks like you meant @gjvis rather than me I think?

darscan commented on README.md:
```diff
@@ -60,6 +60,7 @@ export interface DepGraph {
     version: string | null;
   }>>;
   toJSON(): DepGraphData;
+  isSame(other: DepGraph): boolean;
```
I'd consider a more standard name for this method: `equals()` is what's commonly used in most ecosystems in my experience. Just a thought

Kirill89 replied:
It's not `equals`. Because I'm not comparing root node.

darscan replied:
interesting

Kirill89 replied:
We can also consider to use `equals` with flag `includingRoot = false`.

darscan replied:
I'm curious as to the motivation for excluding the root from comparison. I'm sure there's a good reason, I just can't think of it off the top of my head

Kirill89 replied:
This function initially designed for commit status checks. Just to compare two branches and check if we need to run tests against it.

If it's a library, root node version will change after each release. But we don't care about it because dependencies are same.

michael-go replied:
@Kirill89 nice! didn't about the this. I think it's a good idea to add this `compareRoot: boolean` flag as a param. because in the case of coparing snapshots for example we would like to compare root as swell

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -1,6 +1,7 @@
 import * as _ from 'lodash';
 import * as graphlib from 'graphlib';
 import * as types from './types';
+import {PkgInfo} from './types';
```
probably was added by your IDE :) let's just use `types.PkgInfo`

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +138,10 @@ class DepGraphImpl implements types.DepGraphInternal {
   }
 
+  public isSame(other: types.DepGraph): boolean {
+    return this.isSameNode(this, 'root-node', other as types.DepGraphInternal, 'root-node');
```
please don't assume the root node Id to be `root-node` - it's not enforced. just use `depGraph._rootNodeId` instead 🙏

michael-go commented on test/core/is-same.test.ts:
```diff
@@ -0,0 +1,60 @@
+    expect(a.isSame(b)).toBe(false);
+    expect(b.isSame(a)).toBe(false);
+  });
+});
```
think you are missing a test for the "no comparing root" behavior

Kirill89 commented:
Updates since last review:

- `isSame` -> `equals` CC @darscan 
- `'root-node'` -> `.rootNodeId` CC @michael-go 
- Added optional `compareRoot` flag. `false` by default CC @darscan @michael-go 
- Added tests for  "no comparing root" behaviour (and a lot more tests) CC @michael-go @gjvis

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +138,11 @@ class DepGraphImpl implements types.DepGraphInternal {
     return count;
   }
 
+  public equals(other: types.DepGraph, compareRoot: boolean = false): boolean {
```
would suggest 2 small changes:
* put the `compareRoot` into an options object, this way it will be easier to read & reason about the code that calls this.
* make `compareRoot` true by default, as the root package today is more part of the graph than isn't. e.g the `getPkgs()` returns the root too.

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +138,11 @@ class DepGraphImpl implements types.DepGraphInternal {
   }
 
+  public equals(other: types.DepGraph, compareRoot: boolean = false): boolean {
+    const otherDepGraph = other as types.DepGraphInternal;
```
following @gjvis comments here: https://github.com/snyk/dep-graph/pull/26#discussion_r295833037, suggest instead doing what you started from :) 
```typescript
const otherDepGraph = fromJSON(other.toJSON());
```

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -175,6 +180,66 @@ class DepGraphImpl implements types.DepGraphInternal {
+    compareRoot: boolean,
+    traversedNodes = new Set<string>(),
+  ): boolean {
+    // Skip root nodes comparision.
```
maybe remove or update comment now that it's a param?

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -175,6 +180,66 @@ class DepGraphImpl implements types.DepGraphInternal {
+    const sortFn = (graph: types.DepGraphInternal) => (idA: string, idB: string) => {
+      const pkgA = graph.getNodePkg(idA);
+      const pkgB = graph.getNodePkg(idB);
+      return `${pkgA.name}@${pkgA.version}`.localeCompare(`${pkgB.name}@${pkgB.version}`);
```
note there is already a static utility function `getPkgId(type.Pkg)` for constructing the strings, and it's a bit safer, better use this.

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -175,6 +180,66 @@ class DepGraphImpl implements types.DepGraphInternal {
+    // Compare Each dependency recursively.
+    for (let i = 0; i < depsA.length; i++) {
+      // Prevent cycles.
+      if (!traversedNodes.has(depsA[i])) {
```
Maybe a small bug here? I think you should also make sure the `depsA[i]` == `depsB[i]` - what if A has a cycle, but B doesn't (or points in a different pkg)

Kirill89 replied:
Good catch. I don't want to compare node ids. Will think about possible solution.

Kirill89 commented:
Updated with a separate commit to simplify re-review.

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -175,6 +191,70 @@ class DepGraphImpl implements types.DepGraphInternal {
+    // Compare Each dependency recursively.
+    for (let i = 0; i < depsA.length; i++) {
+      // Prevent cycles.
+      if (traversedNodesA.has(depsA[i]) && traversedNodesB.has(depsB[i])) {
```
@Kirill89 maybe I miss something, but shouldn't we check that the `depsA[i]->package` equals `depsB[i]->package` ? what if both have cycles but with different packages?

Kirill89 replied:
We check it in the recursive call.

Kirill89 replied:
Agree after discussion.

Kirill89 commented on src/core/dep-graph.ts:
```diff
@@ -175,6 +191,70 @@ class DepGraphImpl implements types.DepGraphInternal {
+
+    // Compare Each dependency recursively.
+    for (let i = 0; i < depsA.length; i++) {
+      const pairKey = `${depsA[i]}_${depsB[i]}`;
```
```
      const pairKey = `${depsA[i]}_${depsB[i]}`;

      // Prevent cycles.
      if (traversedPairs.has(pairKey)) {
        continue;
      }

      traversedPairs.add(pairKey);
```

can be rewritten as 

```
      // Prevent cycles.
      if (traversedPairs.size === traversedPairs.add(`${depsA[i]}_${depsB[i]}`).size) {
        continue;
      }
```

but not sure about readability :)

michael-go commented on src/core/dep-graph.ts:
```diff
@@ -138,6 +139,21 @@ class DepGraphImpl implements types.DepGraphInternal {
+      otherDepGraph = other as types.DepGraphInternal;
+    } else {
+      // At runtime theoretically we can have multiple versions of
+      // @snyk/dep-graph. If "other" in not an instance of the same class it is
```
typo: If "other" ~in~ is not  an

snyksec commented:
:tada: This PR is included in version 1.8.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@snyk/dep-graph)
- [GitHub release](https://github.com/snyk/dep-graph/releases/tag/v1.8.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
